File: ProjectEuler/P033.cs
using System;

namespace ProjectEuler
{
    // Find the value of the denominator of the product of four non-trivial fractions
    public class P033
    {
        int nom = 1, den = 1;

        int GCD(int a, int b)
        {
            return b == 0 ? a : GCD(b, a % b);
        }

        public void Check(double ans, char a, char b)
        {
            var x = Char.GetNumericValue(a);
            var y = Char.GetNumericValue(b);
            if (ans == x / y)
            {
                nom *= (int)x;
                den *= (int)y;
            }
        }

        // brute force will easily take less than 1ms
        public int Solve()
        {
            const int n = 100;

            for (int i = 11; i < n; i++)
            {
                for (int j = i + 1; j < n; j++)
                {
                    if (i % 10 == 0 || j % 10 == 0)
                    {
                        continue;
                    }
                    
                    var ans = (double)i / j;
                    string a = i.ToString(), b = j.ToString();

                    if (a[0] == b[0])
                    {
                        Check(ans, a[1], b[1]);
                    }
                    else if (a[0] == b[1])
                    {
                        Check(ans, a[0], b[1]);
                    }
                    else if (a[1] == b[0])
                    {
                        Check(ans, a[0], b[1]);
                    }
                    else if (a[1] == b[1])
                    {
                        Check(ans, a[0], b[0]);
                    }
                }
            }

            return den / GCD(nom, den);
        }
    }
}
        

File: ProjectEuler/Problems 1 - 50/P033.py
# https://projecteuler.net/problem=33
# Find the value of the denominator of the product
# of four non-trivial fractions

import fractions as f

result = 1
n = 100


def check(a, b, i, j):
    global result
    x = f.Fraction(int(a), int(b))
    y = f.Fraction(i, j)
    if x == y:
        result *= x

for i in range(11, n):
    for j in range(i + 1, n):
        if not i % 10 or not j % 10:
            continue

        a, b = str(i), str(j)

        if a[0] == b[0]:
            check(a[1], b[1], i, j)
        elif a[0] == b[1]:
            check(a[0], b[1], i, j)
        elif a[1] == b[0]:
            check(a[0], b[1], i, j)
        elif a[1] == b[1]:
            check(a[0], b[0], i, j)

print(result.denominator)
